daemon/graphdriver/copy/copy.go

```diff
@@ -107,156 +119,395 @@ func copyXattr(srcPath, dstPath, attr string) error {
+	c := w.workers[w.rand.Intn(len(w.workers))]
+
+	names, err := readDirNames(path)
+	err1 := w.walkFunc(path, info, c, err)
+	// If err != nil, walk can't walk into this directory.
+	// err1 != nil means walkFn want walk to skip this directory or stop walking.
+	// Therefore, if one of err and err1 isn't nil, walk will return.
+	if err != nil || err1 != nil {
+		// The caller's behavior is controlled by the return value, which is decided
+		// by walkFn. walkFn may ignore err and return nil.
+		// If walkFn returns SkipDir, it will be handled by the caller.
+		// So walk should return whatever walkFn returns.
+		return err1
+	}
+
+	for _, name := range names {
+		filename := filepath.Join(path, name)
+		fileInfo, err := os.Lstat(filename)
+		if err != nil {
+			if err := w.walkFunc(filename, fileInfo, c, err); err != nil {
 				return err
 			}
-		}
-
-		// system.Chtimes doesn't support a NOFOLLOW flag atm
-		// nolint: unconvert
-		if f.IsDir() {
-			dirsToSetMtimes.PushFront(&dirMtimeInfo{dstPath: &dstPath, stat: stat})
-		} else if !isSymlink {
-			aTime := time.Unix(int64(stat.Atim.Sec), int64(stat.Atim.Nsec))
-			mTime := time.Unix(int64(stat.Mtim.Sec), int64(stat.Mtim.Nsec))
-			if err := system.Chtimes(dstPath, aTime, mTime); err != nil {
+		} else if fileInfo.IsDir() {
+			err = w.walk2(filename, fileInfo)
+			if err != nil {
 				return err
 			}
 		} else {
-			ts := []syscall.Timespec{stat.Atim, stat.Mtim}
-			if err := system.LUtimesNano(dstPath, ts); err != nil {
+			if err := w.walkFunc(filename, fileInfo, c, err); err != nil {
 				return err
 			}
 		}
-		return nil
-	})
+	}
+	return w.updateDirTime(path, info, c)
+}
+
+func (w *walkFn) walk(root string) error {
+	c := w.workers[w.rand.Intn(len(w.workers))]
+	info, err := os.Lstat(root)
 	if err != nil {
+		err = w.walkFunc(root, nil, c, err)
+	} else {
+		err = w.walk2(root, info)
+	}
+	return err
+}
+
+// DirCopy copies or hardlinks the contents of one directory to another,
+// properly handling xattrs, and soft links
+//
+// Copying xattrs can be opted out of by passing false for copyXattrs.
+func DirCopy(srcDir, dstDir string, copyMode Mode, copyXattrs bool) error {
+	return DirCopyWithConcurrency(srcDir, dstDir, copyMode, copyXattrs, 0)
+}
+
+// DirCopyWithConcurrency performs the same work as DirCopy, but allows you to specify the
+// concurrency level. Specifying workerCount as 0 will make it use the default worker
+// count
+func DirCopyWithConcurrency(srcDir, dstDir string, copyMode Mode, copyXattrs bool, workerCount int) error {
+	if workerCount < 0 {
+		return fmt.Errorf("Copy concurrency '%d' is less than 0", workerCount)
+	}
+	if workerCount == 0 {
+		workerCount = defaultWorkerCount
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	errGroup, errGroupCtx := errgroup.WithContext(ctx)
+	workers := make([]chan *fileCopyInfo, workerCount)
+	for i := 0; i < len(workers); i++ {
+		c := make(chan *fileCopyInfo, 100)
```

Comment:
100 should be a named `const`. (no need to make this `storage-opt`, probably 😄 )